pkg/webhookconfig/registration.go

```diff
@@ -63,6 +65,10 @@ func (wrc *Register) Register() error {
 	if wrc.serverIP != "" {
 		logger.Info("Registering webhook", "url", fmt.Sprintf("https://%s", wrc.serverIP))
 	}
+	if err := wrc.checkEndpoint(); err != nil {
```

Comment:
@treydock - thanks for sending the fix!

This is a very interesting finding and I'm thinking whether we can combine the liveness probe with the webhook registration - currently, Kyverno has a webhook monitor that runs every 30s to verify the webhooks are healthy and set the annotation `kyverno.io/webhookActive` in the deployment accordingly. So it's possible that Kyverno Pod is running and ready but it's actually not serving any admission requests. Having this check in the livenessProbe forces the restart of the container, thus notifying the user that something goes wrong with the webhook. I feel this would be a nicer approach than what we have now, I would love to know your thoughts on this. If it sounds reasonable, I can create another issue for it.

To fix your issue (with the endpoint), this change looks good. Just that we need to add an infinite loop (with a timeout or failure threshold) to check the endpoint and register the webhook accordingly in the [main tread](https://github.com/kyverno/kyverno/blob/main/cmd/kyverno/main.go#L300-L303). Otherwise if this check fails, the webhook monitor will need to wait for another 30s in order to register the webhooks. Let me know if you are interested in sending the fix, otherwise I can pick this up.

Reply:
The liveness idea is interesting, though I'm not sure it would help my particular issue since the pod itself was healthy just the service endpoint was not yet ready and restarting the pod if endpoint is not ready won't help get endpoint ready since pod gets restarted, if I understand how Kubernetes determines when to assign an endpoint IP to a service.

I'm not sure I follow the infinite loop idea. Are you talking about some check inside the main monitor loop to avoid a never ending continuation of endpoint never being ready?  If so not sure what that would look like, maybe an error counter for how many times registering the webhook failed?  When I tested this locally and I think one reason the e2e tests failed initially is there was actually only about a 5s delay getting the webhooks added when the endpoint wasn't initially ready.  I don't recall things having to wait the full 30s.  So from that main thread what are you envisioning, maybe checking the endpoint in some timed loop in main thread?

Reply:
@realshuting  I tried to add some logic to main to wait for webhook to become ready but that failed badly because the pod was never seen as ready so Kubernetes never setup the service endpoint.  Any pointers on how to go about solving that problem are welcome, or if preferred to handle that yourself, either way works for me.

FWIW this is what I added 
```
	registerTimeout := time.After(30 * time.Second)
	registerTicker := time.NewTicker(5 * time.Second)
	for {
		err = webhookCfg.Register()
		if err != nil {
			setupLog.Error(err, "Failed to register admission control webhooks")
		} else {
			break
		}
		select {
		case <-registerTimeout:
			setupLog.Error(err, "Timeout registering admission control webhooks")
			os.Exit(1)
		case <-registerTicker.C:
			err = webhookCfg.Register()
			if err != nil {
				setupLog.Error(err, "Failed to register admission control webhooks")
			} else {
				break
			}
		}
	}
	registerTicker.Stop()
```

Reply:
>So from that main thread what are you envisioning, maybe checking the endpoint in some timed loop in main thread?

Yes this was the idea, but we have to make it a goroutine because the webhook server starts after this (that was the reason the Pod never turned ready). And we can either make the ticker with shorter interval (1s) or use `ExponentialBackoff` ([example](https://github.com/kyverno/kyverno/blob/ca5a4e1986d08b174fec796f6ba448d6d379e0d9/pkg/policyreport/common.go#L27-L34)) for this check.

Reply:
I added a commit that I think fixes the issue in main.  The logging is kind of verbose and maybe something needing to be improved but based on logs the webhooks got added after an about 5 second delay:

```
E0330 14:12:27.256607       1 registration.go:502] Register "msg"="ns" "error"="Endpoint not ready"  "kyverno"="name" "kyverno-svc"=null
I0330 14:12:27.256650       1 registration.go:69] Register "msg"="Endpoint not ready, skipping registration of webhook"  "err"="Endpoint not ready"
E0330 14:12:27.256678       1 main.go:311] setup "msg"="Failed to register admission control webhooks" "error"="Endpoint not ready"  
E0330 14:12:28.256539       1 registration.go:502] Register "msg"="ns" "error"="Endpoint not ready"  "kyverno"="name" "kyverno-svc"=null
I0330 14:12:28.256576       1 registration.go:69] Register "msg"="Endpoint not ready, skipping registration of webhook"  "err"="Endpoint not ready"
E0330 14:12:28.256587       1 main.go:311] setup "msg"="Failed to register admission control webhooks" "error"="Endpoint not ready"  
E0330 14:12:29.257372       1 registration.go:502] Register "msg"="ns" "error"="Endpoint not ready"  "kyverno"="name" "kyverno-svc"=null
I0330 14:12:29.257407       1 registration.go:69] Register "msg"="Endpoint not ready, skipping registration of webhook"  "err"="Endpoint not ready"
E0330 14:12:29.257418       1 main.go:311] setup "msg"="Failed to register admission control webhooks" "error"="Endpoint not ready"  
E0330 14:12:30.256088       1 registration.go:502] Register "msg"="ns" "error"="Endpoint not ready"  "kyverno"="name" "kyverno-svc"=null
I0330 14:12:30.256123       1 registration.go:69] Register "msg"="Endpoint not ready, skipping registration of webhook"  "err"="Endpoint not ready"
E0330 14:12:30.256134       1 main.go:311] setup "msg"="Failed to register admission control webhooks" "error"="Endpoint not ready"  
E0330 14:12:31.257325       1 registration.go:502] Register "msg"="ns" "error"="Endpoint not ready"  "kyverno"="name" "kyverno-svc"=null
I0330 14:12:31.257366       1 registration.go:69] Register "msg"="Endpoint not ready, skipping registration of webhook"  "err"="Endpoint not ready"
E0330 14:12:31.257379       1 main.go:311] setup "msg"="Failed to register admission control webhooks" "error"="Endpoint not ready"  
E0330 14:12:32.256497       1 registration.go:502] Register "msg"="ns" "error"="Endpoint not ready"  "kyverno"="name" "kyverno-svc"=null
I0330 14:12:32.256532       1 registration.go:69] Register "msg"="Endpoint not ready, skipping registration of webhook"  "err"="Endpoint not ready"
E0330 14:12:32.256543       1 main.go:311] setup "msg"="Failed to register admission control webhooks" "error"="Endpoint not ready"  
I0330 14:12:33.258030       1 registration.go:497] Register "msg"="Endpoint ready"  "name"="kyverno-svc" "ns"="kyverno"
I0330 14:12:33.258070       1 registration.go:297] Register "msg"="deleting all webhook configurations"  
I0330 14:12:33.278102       1 registration.go:291] Register "msg"="created webhook"  "kind"="MutatingWebhookConfiguration" "name"="kyverno-verify-mutating-webhook-cfg"
I0330 14:12:33.292101       1 registration.go:246] Register "msg"="created webhook"  "kind"="ValidatingWebhookConfiguration" "name"="kyverno-policy-validating-webhook-cfg"
I0330 14:12:33.301111       1 registration.go:269] Register "msg"="created webhook"  "kind"="MutatingWebhookConfiguration" "name"="kyverno-policy-mutating-webhook-cfg"
I0330 14:12:33.458275       1 registration.go:223] Register "msg"="created webhook" "kind"="ValidatingWebhookConfiguration" "name"="kyverno-resource-validating-webhook-cfg" 
I0330 14:12:33.657374       1 registration.go:197] Register "msg"="created webhook" "kind"="MutatingWebhookConfiguration" "name"="kyverno-resource-mutating-webhook-cfg" 
I0330 14:12:56.863032       1 status.go:70] WebhookMonitor/WebhookStatusControl "msg"="updating deployment annotation" "name"="kyverno" "namespace"="kyverno" "key"="kyverno.io/webhookActive" "val"="true"
```

Reply:
Added commit to reduce duplicate log entries and ensure registration returns error.